azurelinuxagent/common/protocol/extensions_config_retriever.py

```diff
@@ -0,0 +1,348 @@
+                self._set_fabric(self._last_incarnation, self._last_svd_seqNo)
+
+    def _reset(self):
+        """
+        Removes all cache files and resets all cached goal state information
+        This is necessary if a VM image is deployed from this one so we start fresh
+        """
+        self._remove_cache(INCARNATION_FILE_NAME)
+        self._remove_cache(SEQUENCE_NUMBER_FILE_NAME)
+        self._remove_cache(SVD_SEQNO_FILE_NAME)
+        self._remove_cache(GOAL_STATE_SOURCE_FILE_NAME)
+        self._remove_cache(VM_ID_FILE_NAME)
+
+    def _reset_if_necessary(self):
+        cached_vm_id = self._get_cached_vm_id()
+        current_vm_id = self._get_vm_id()
+        if current_vm_id is None:
+            logger.warn("Unable to retrieve the current vm id. Skipping reset")
+        elif cached_vm_id is None:
+            logger.info("Remembering current vm id is {0}".format(current_vm_id))
+            self._set_cached_vm_id(current_vm_id)
+        elif current_vm_id != cached_vm_id:
+            logger.warn("The vm id has changed from {0} to {1}. Resetting cached state".format(cached_vm_id, current_vm_id))
+            self._reset()
+            self._set_cached_vm_id(current_vm_id)
+
+    def _remove_extensions_if_necessary(self, extensions_config):
+        """
+        If this is a Fabric GS, but the InSvdSeqNo did NOT change, then the goal state was
+        created directly by Fabric and bypassed CRP. A common scenario is remote access.
+        Another is when wire server restarts and uses a new incarnation.
+        The problem is any extensions contained here may be out of date, because they were
+        more recently updated via a FastTrack GS. Therefore, we remove them here in that case.
+        """
+        if self._pending_mode == GOAL_STATE_SOURCE_FABRIC:
+            svd_seqNo = self._get_svd_seqNo()
+            if str(extensions_config.svd_seqNo) == str(svd_seqNo):
+                logger.info("SvdSeqNo did not change. Removing extensions from goal state")
+                extensions_config.ext_handlers = None
+                self._last_svd_seqNo = extensions_config.svd_seqNo
+                self._pending_svd_seqNo = extensions_config.svd_seqNo
+            else:
+                self._pending_svd_seqNo = svd_seqNo
+
+    def _decide_what_to_process(self, fabric_changed, fast_track_changed):
+        """
+        If just Fabric GS changed, then process only that.
+        If just FastTrack GS changed, then process only that.
+        If both changed, then process Fabric and then FastTrack.
+        If neither changed, then process whichever we used last (to keep with the current behavior)
+        """
+        if fabric_changed:
+            self._set_reason("FabricChanged")
+            return GOAL_STATE_SOURCE_FABRIC
+        if fast_track_changed:
+            self._set_reason("FTChanged")
+            return GOAL_STATE_SOURCE_FASTTRACK
+        if self._get_mode() == GOAL_STATE_SOURCE_FASTTRACK:
```

Comment:
We should first check if `self._last_mode` is None before trying to read from File

Reply:
changed